contributors/design-proposals/stable-pods.md

```diff
@@ -0,0 +1,93 @@
+orchestrators running on different machines such as the ReplicaSet or the
+PodDisruptionBudget controller will merely need to look at the Stable condition that is set
+in the status of a Pod.
+
+Services should not be affected by the implementation of Stable Pods. Endpoints should
+continue to be added or removed when a Pod transitions to or from Ready.
+
+### API changes
+
+A new field is proposed in the PodSpec:
+```go
+	// Minimum number of seconds for which a newly created Pod should be Ready
+	// without any of its container crashing, for it to be considered Stable.
+	// Defaults to 0 (pod will be considered Stable as soon as it is Ready)
+	// +optional
+	MinReadySeconds *int32 `json:"minReadySeconds,omitempty"`
+```
+and a new PodConditionType:
+```go
+	// PodStable is added in a Pod after it has been Ready for at least MinReadySeconds.
+	// The Pod should already have an Endpoint and serve requests, this condition lets
+	// higher-level orchestrators know that the Pod is stable after some amount of time
+	// without having any of its containers crashed.
+	PodStable PodConditionType = "Stable"
+```
+
+The reason for adding a separate Condition instead of extending the existing Ready
+condition is mainly for allowing traffic in Pods. Users can have robust rollouts if their
+applications expose metrics which are then fed into readiness checks.
+
+Additionally:
+* Deployments/ReplicaSets/DaemonSets already use MinReadySeconds in their Spec so we
+should probably deprecate those fields in favor of the field in the PodTemplateSpec and
+remove them in a future version. 
+* All the aforementioned controllers will not propagate MinReadySeconds from their Spec
+down to their PodTemplateSpec because that may lead in differences in the PodTemplateSpec
+between a Deployment and a ReplicaSet resulting in new rollouts. If MinReadySeconds is
+specified both in the Spec and PodTemplateSpec for a controller, the field in the
+PodTemplateSpec will stomp the field in the Spec. If it is specified only in the Spec,
+Pods can be created using it without updating the controller's PodTemplateSpec.
+
+### kubelet changes
+For a Pod that specifies MinReadySeconds, kubelet will need to rerun the readiness check
+after MinReadySeconds and also check if any of its containers has crashed in the meantime.
+If everything is ok, the kubelet will switch the Pod's Stable condition to Status=True.
+Pods that don't specify MinReadySeconds will be considered Stable as soon as they are
+Ready.
+
+### Controller manager changes
+The ReplicaSet and DaemonSet controllers will create new Pods by setting MinReadySeconds in
+the PodSpec if it is specified either in their Spec or PodTemplateSpec. The value in the
+PodTemplateSpec should stomp the value in the Spec. For Pods that do not specify
+MinReadySeconds (those running on old kubelets), the controllers will continue to use the
+current approach for estimating stability. Eventually, we should switch all controllers that
```

Comment:
Can we make PodTemplate.Spec.MinReadySeconds default to whatever is set in the controller spec when creating pods (we don't need to touch PodTemplate so won't cause additional rollouts)? It overwrites controller Spec.MinReadySeconds only when it's set (not `nil`).